### ISSUE TITLE
Add new CRD TemporalWorker to manage worker fleets

### ISSUE THREAD

**Issue by ktenzer** [alexandrevilain/temporal-operator#149]
It would be great to have a CRD endpoint to manage workers. The temporal-operator could then be used not only by those self-hosting but also to help scale and manage workers which would likely be controlled by another group (developers instead of platform engineers). 

Initial ideas:
- CR should control worker farm for a given app/workflow
- CR should allow setting resource limits
- CR should support worker mtls and handling cert bootstrap via ClusterClient CR
- CR should allow image to be selected
- CR should perform upgrade strategy if the image is changed
- CR should have clear naming convention and differentiate with internal worker(s) deployed by temporal server CR

Nice to have:
- CR to support autoscaling using server metrics to detect worker starvation
- CR to support code injection through mount or maybe sidecar
- CR to support SDK metrics metering
- CR to support CI through tekton or something else

**Comment by ktenzer** [comment]
@alexandrevilain definitely looking for your feedback on this

**Comment by alexandrevilain** [comment]
Hi @ktenzer 

What a great idea! I think it could become a killer feature for our users! Thanks for this proposition! 

At first glance, I see few things we should care about before starting the development:

- The naming. I often see confusion between the Worker service which is part of the temporal cluster and workers the user should implement. So we have to take care about naming and documentation.
- The "locking". As we're taking opinion in the way of deploying worker, we should ensure that users will not have to do some big code changes to start using this CRD to deploy their workers. We will probably have to support volumes mounts, env variable or even sidecar injection to provide endpoint, credentials, certificates to the worker.
- We have to still care about "classic" worker deployment. And if the user wants mTLS he should be able to use the TemporalClusterClient to retrieve its mTLS credentials.

WDYT ?

**Comment by ktenzer** [comment]
> Hi @ktenzer
> 
> What a great idea! I think it could become a killer feature for our users! Thanks for this proposition!
> 
> At first glance, I see few things we should care about before starting the development:
> 
> * The naming. I often see confusion between the Worker service which is part of the temporal cluster and workers the user should implement. So we have to take care about naming and documentation.
> * The "locking". As we're taking opinion in the way of deploying worker, we should ensure that users will not have to do some big code changes to start using this CRD to deploy their workers. We will probably have to support volumes mounts, env variable or even sidecar injection to provide endpoint, credentials, certificates to the worker.
> * We have to still care about "classic" worker deployment. And if the user wants mTLS he should be able to use the TemporalClusterClient to retrieve its mTLS credentials.
> 
> WDYT ?

I will add these points to requirements above.

Agree naming convention is critical. In the end the worker deployed w/temporal server is just another worker but it is used for orchestrating internal activities of Temporal.

Initially I would propose we simply deploy an image provided by the developer but I think we should eventually add volume mounts and injection. For v1 idea would be developer CI/CD produces container image as artifact and this would then be consumed by CR. Eventually though as you mentioned it would be really nice to inject code via mount or maybe sidecar.

Regarding mTLS yes this would be something we need to enable in v1 so that a worker can leverage the ClusterClient. Probably we can just create that CR but user would have to grab those certs from pre-defined ENV vars available to worker but I assume that should all work with what we have today.

**Comment by alexandrevilain** [comment]
Hi ! 
As seen with you, first iteration for this feature will ship with release 0.8.0. 
I purpose to keep this issue open for incoming features:
- default builder using kaniko
- builder jobs customization options
- autoscaling

**Comment by alexandrevilain** [comment]
Closing in favor of: https://github.com/alexandrevilain/temporal-operator/issues/205
